server/tidb_test.go

```diff
@@ -131,3 +132,14 @@ func (ts *TidbTestSuite) TestSocket(c *C) {
 	dsn = tcpDsn
 	server.Close()
 }
+
+func (ts *TidbTestSuite) TestIssue3662(c *C) {
+	db, _ := sql.Open("mysql", "root@tcp(localhost:4001)/a_database_not_exist")
+	defer db.Close()
+	// According to documentation, "Open may just validate its arguments without
+	// creating a connection to the database. To verify that the data source name
+	// is valid, call Ping."
+	err := db.Ping()
```

Comment:
actually `Open()` itself does not create connection immediately as explained in the comment. From my testing, when using a single `Open()` statement (without other queries or `Ping`), no errors will be thrown even if the authentication or handshake should fail.